tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Signature.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright 2020 The TensorFlow Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.tensorflow;
+
+import java.util.Map;
+import java.util.Set;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.proto.framework.DataType;
+import org.tensorflow.proto.framework.SignatureDef;
+import org.tensorflow.proto.framework.TensorInfo;
+import org.tensorflow.proto.framework.TensorShapeProto;
+import org.tensorflow.proto.framework.TensorShapeProto.Dim;
+
+/**
+ * Describe the inputs and outputs of an executable entity, such as a {@link ConcreteFunction}, among
+ * other useful metadata.
+ */
+public class Signature  {
+
+  /** The default signature key, when not provided */
+  public static final String DEFAULT_KEY = "serving_default";
+
+  /**
+   * Builds a new function signature.
+   */
+  public static class Builder {
+
+    /**
+     * Sets the unique key of this signature.
+     *
+     * <p/>When not set explicitly, the default value is {@link #DEFAULT_KEY}.
+     *
+     * @param key signature key
+     * @return this builder
+     */
+    public Builder key(String key) {
+      this.key = key;
+      return this;
+    }
+
+    /**
+     * Register a tensor as an input of the function.
+     *
+     * @param inputName user-friendly name for this input tensor
+     * @param input input tensor
+     * @return this builder
+     */
+    public Builder input(String inputName, Operand<?> input) {
+      signatureBuilder.putInputs(inputName, toTensorInfo(input.asOutput()));
+      return this;
+    }
+
+    /**
+     * Register a tensor as an output of the function.
+     *
+     * @param inputName user-friendly name for this input tensor
+     * @param input input tensor
+     * @return this builder
+     */
+    public Builder output(String outputName, Operand<?> output) {
+      signatureBuilder.putOutputs(outputName, toTensorInfo(output.asOutput()));
+      return this;
+    }
+
+    /**
+     * Provide extensible name information enabling third-party users to mark a signature as
+     * supporting a particular method
+     *
+     * @param methodName method name
+     * @return this builder
+     */
+    public Builder methodName(String methodName) {
+      signatureBuilder.setMethodName(methodName);
```

Comment:
Similarly should we check it isn't empty or null?

Reply:
I'll check but in this case, it is allowed to have no method name assigned to a signature, hence `null` should be accepted as a value. I can document this.